can/interfaces/etas/boa.py

```diff
@@ -0,0 +1,672 @@
+import ctypes
+
+try:
+    # try to load libraries from the system default paths
+    _csi = ctypes.windll.LoadLibrary("dll-csiBind")
+    _oci = ctypes.windll.LoadLibrary("dll-ocdProxy")
+except FileNotFoundError:
+    # try to load libraries with hardcoded paths
+    if ctypes.sizeof(ctypes.c_voidp) == 4:
+        # 32 bit
+        path = "C:/Program Files (x86)/ETAS/BOA_V2/Bin/Win32/Dll/Framework/"
+    elif ctypes.sizeof(ctypes.c_voidp) == 8:
+        # 64 bit
+        path = "C:/Program Files/ETAS/BOA_V2/Bin/x64/Dll/Framework/"
+    _csi = ctypes.windll.LoadLibrary(path + "dll-csiBind")
+    _oci = ctypes.windll.LoadLibrary(path + "dll-ocdProxy")
+
+# Common (BOA)
+
+BOA_ResultCode = ctypes.c_uint32
+BOA_Handle = ctypes.c_int32
+BOA_Time = ctypes.c_int64
+
+BOA_NO_VALUE = -1
+BOA_NO_HANDLE = BOA_Handle(BOA_NO_VALUE)
+BOA_NO_TIME = BOA_Time(BOA_NO_VALUE)
+
+
+class BOA_UuidBin(ctypes.Structure):
+    _fields_ = [("data", ctypes.c_uint8 * 16)]
+
+
+class BOA_Version(ctypes.Structure):
+    _fields_ = [
+        ("majorVersion", ctypes.c_uint8),
+        ("minorVersion", ctypes.c_uint8),
+        ("bugfix", ctypes.c_uint8),
+        ("build", ctypes.c_uint8),
+    ]
+
+
+class BOA_UuidVersion(ctypes.Structure):
+    _fields_ = [("uuid", BOA_UuidBin), ("version", BOA_Version)]
+
+
+class BOA_ServiceId(ctypes.Structure):
+    _fields_ = [("api", BOA_UuidVersion), ("access", BOA_UuidVersion)]
+
+
+class BOA_ServiceIdParam(ctypes.Structure):
+    _fields_ = [
+        ("id", BOA_ServiceId),
+        ("count", ctypes.c_uint32),
+        ("accessParam", ctypes.c_char * 128),
+    ]
+
+
+# Connection Service Interface (CSI)
+
+# CSI - Search For Service (SFS)
+
+CSI_NodeType = ctypes.c_uint32
+CSI_NODE_MIN = CSI_NodeType(0)
+CSI_NODE_MAX = CSI_NodeType(0x7FFF)
+
+
+class CSI_NodeRange(ctypes.Structure):
+    _fields_ = [("min", CSI_NodeType), ("max", CSI_NodeType)]
+
+
+class CSI_SubItem(ctypes.Structure):
+    _fields_ = [
+        ("server", BOA_ServiceIdParam),
+        ("nodeType", CSI_NodeType),
+        ("uriName", ctypes.c_char * 128),
+        ("visibleName", ctypes.c_char * 4),
+        ("version", BOA_Version),
+        ("reserved2", ctypes.c_char * 88),
+        ("serverAffinity", BOA_UuidBin),
+        ("requiredAffinity0", BOA_UuidBin),
+        ("reserved", ctypes.c_int32 * 4),
+        ("requiredAffinity1", BOA_UuidBin),
+        ("count", ctypes.c_int32),
+        ("requiredAPI", BOA_ServiceId * 4),
+    ]
+
+
+class CSI_Tree(ctypes.Structure):
+    pass
+
+
+CSI_Tree._fields_ = [
+    ("item", CSI_SubItem),
+    ("sibling", ctypes.POINTER(CSI_Tree)),
+    ("child", ctypes.POINTER(CSI_Tree)),
+    ("childrenProbed", ctypes.c_int),
+]
+
+CSI_CreateProtocolTree = _csi.CSI_CreateProtocolTree
+CSI_CreateProtocolTree.argtypes = [
+    ctypes.c_char_p,
+    CSI_NodeRange,
+    ctypes.POINTER(ctypes.POINTER(CSI_Tree)),
+]
+CSI_CreateProtocolTree.restype = BOA_ResultCode
```

Comment:
One could also add `CSI_CreateProtocolTree.errcheck = ...` to use [this](https://docs.python.org/3/library/ctypes.html#ctypes._FuncPtr.errcheck) (and similarly for the other methods). Just an idea, maybe it simplifies the bus class?